core/src/main/java/com/linecorp/armeria/server/HttpService.java

```diff
@@ -57,4 +59,15 @@ R decorate(Function<? super HttpService, R> decorator) {
     default HttpService decorate(DecoratingHttpServiceFunction function) {
         return new FunctionalDecoratingHttpService(this, function);
     }
+
+    /**
+     * Determines an {@link ExchangeType} for this {@link HttpService} from the given {@link RequestHeaders}
+     * and {@link Route}. By default, {@link ExchangeType#BIDI_STREAMING} is set.
+     *
+     * <p>Note that an {@link HttpRequest} will be aggregated before serving the {@link HttpService} if
+     * {@link ExchangeType#UNARY} or {@link ExchangeType#RESPONSE_STREAMING} is set.
+     */
+    default ExchangeType exchangeType(RequestHeaders headers, Route route) {
```

Comment:
QQ: We can't use `ServiceRequestContext` here because it's too early, right?

Reply:
It was. Come to think of it, it might be possible to create `ServiceRequestContext` by more moving code in `HttpServerHandler` to `Http*Request`.

Reply:
I just realized that we can create `ServiceRequestContext` here. 😅
Because the implementation of `DecodedHttpRequest` will be chosen depending on the result of `HttpService.exchangeType(...)`.
As you know, an `HttpRequest` will be required to create a `ServiceRequestContext`.

Reply:
>  we can create ServiceRequestContext here. 

Oops.. we can't create ServiceRequestContext here.